From df22ffa9f1ab2616c92ad8f6a0ea2983b78407ad Mon Sep 17 00:00:00 2001
From: Alex Rocha <rochaa@tacc.utexas.edu>
Date: Thu, 10 Jan 2019 16:54:02 -0600
Subject: [PATCH 49/52] Getting Chameleon portal logout url from settings,
 setting up cc sso so that it can be disabled via settings

---
 horizon/templates/auth/_logout_form.html |  3 +--
 openstack_dashboard/cc_web_sso_urls.py   | 39 +++++++++++++++++---------------
 openstack_dashboard/cc_websso_views.py   | 15 +++++++-----
 openstack_dashboard/settings.py          |  2 +-
 4 files changed, 32 insertions(+), 27 deletions(-)

diff --git a/horizon/templates/auth/_logout_form.html b/horizon/templates/auth/_logout_form.html
index c2e35e2..a35ebce 100644
--- a/horizon/templates/auth/_logout_form.html
+++ b/horizon/templates/auth/_logout_form.html
@@ -27,9 +27,8 @@
 
     <div class="panel-body">
         <span>You may still be logged in to Chameleon Portal, to sign out of the portal, please go to the sign out page </span>
-        <a href="www.chameleoncloud.org/logout">here</a>.
+        <a href="{{cc_logout_url}}">here</a>.
     </div>
-
     <div class="panel-footer">
       {% block login_footer %}
         <button id="loginBtn" type="submit" class="btn btn-primary pull-right">
diff --git a/openstack_dashboard/cc_web_sso_urls.py b/openstack_dashboard/cc_web_sso_urls.py
index 7119fc1..fe0cace 100644
--- a/openstack_dashboard/cc_web_sso_urls.py
+++ b/openstack_dashboard/cc_web_sso_urls.py
@@ -1,26 +1,29 @@
 from django.conf.urls import url
-
+from django.conf import settings
 from openstack_auth import utils
 from openstack_auth import views
 from openstack_dashboard import cc_websso_views
 utils.patch_middleware_get_user()
 
+cc_websso_enabled = getattr(settings, 'CC_WEBSSO_ENABLED', False)
+if(cc_websso_enabled):
+    urlpatterns = [
+        url(r"^login/$", cc_websso_views.login, name='login'),
+        url(r"^logout/$", cc_websso_views.logout, name='logout'),
+        url(r'^switch/(?P<tenant_id>[^/]+)/$', views.switch,
+            name='switch_tenants'),
+        url(r'^switch_services_region/(?P<region_name>[^/]+)/$',
+            views.switch_region,
+            name='switch_services_region'),
+        url(r'^switch_keystone_provider/(?P<keystone_provider>[^/]+)/$',
+            views.switch_keystone_provider,
+            name='switch_keystone_provider')
+    ]
 
-urlpatterns = [
-    url(r"^login/$", cc_websso_views.login, name='login'),
-    url(r"^logout/$", cc_websso_views.logout, name='logout'),
-    url(r'^switch/(?P<tenant_id>[^/]+)/$', views.switch,
-        name='switch_tenants'),
-    url(r'^switch_services_region/(?P<region_name>[^/]+)/$',
-        views.switch_region,
-        name='switch_services_region'),
-    url(r'^switch_keystone_provider/(?P<keystone_provider>[^/]+)/$',
-        views.switch_keystone_provider,
-        name='switch_keystone_provider')
-]
-
-if utils.is_websso_enabled():
-    urlpatterns.append(url(r"^websso/$", views.websso, name='websso'))
+    if utils.is_websso_enabled():
+        urlpatterns.append(url(r"^websso/$", views.websso, name='websso'))
 
-if utils.is_websso_enabled():
-    urlpatterns.append(url(r"^ccwebsso/$", cc_websso_views.cc_websso, name='custom_websso'))
\ No newline at end of file
+    if utils.is_websso_enabled():
+        urlpatterns.append(url(r"^ccwebsso/$", cc_websso_views.cc_websso, name='custom_websso'))
+else:
+    urlpatterns = []
\ No newline at end of file
diff --git a/openstack_dashboard/cc_websso_views.py b/openstack_dashboard/cc_websso_views.py
index 1deea3e..a0af0e2 100644
--- a/openstack_dashboard/cc_websso_views.py
+++ b/openstack_dashboard/cc_websso_views.py
@@ -10,7 +10,6 @@ from django.views.decorators.debug import sensitive_post_parameters
 from openstack_auth import utils
 from openstack_auth import exceptions
 from openstack_auth import user as auth_user
-from django.conf import settings
 import six
 
 LOG = logging.getLogger(__name__)
@@ -21,15 +20,16 @@ LOG = logging.getLogger(__name__)
 def login(request, template_name=None, extra_context=None, **kwargs):
     if not request.user.is_authenticated():
         host = getattr(settings, 'SSO_CALLBACK_HOST', None)
-        cc_portal_url = getattr(settings, 'CHAMELEON_PORTAL_SSO_URL', None)
+        cc_portal_url = getattr(settings, 'CHAMELEON_PORTAL_SSO_BASE_URL', None) + getattr(settings, 'CHAMELEON_PORTAL_SSO_LOGIN_PATH', None)
         if(host is None or cc_portal_url is None):
             LOG.error('Misconfigured CC Portal SSO, settings:, '
                 + 'CHAMELEON_PORTAL_SSO_URL: ' + str(cc_portal_url) + ', SSO_CALLBACK_HOST: ' + str(host))
             raise Exception('SSO Login Error')
-        next = ''
-        if request.GET.get('next'):
-            next = '&next=' + request.GET.get('next')
         login_url = cc_portal_url + '?host=' + host
+        if request.GET.get('next'):
+            login_url += '&next=' + request.GET.get('next')
+        if getattr(settings, 'WEBROOT', None) and getattr(settings, 'WEBROOT', None) != '/':
+            login_url += '&webroot=' + getattr(settings, 'WEBROOT', '')
         return django_http.HttpResponseRedirect(login_url)
     return openstack_auth.views.login(request, template_name=None, extra_context=None, **kwargs)
 
@@ -77,6 +77,9 @@ def logout(request, login_url=None, **kwargs):
     msg = 'Logging out user "%(username)s".' % \
         {'username': request.user.username}
     LOG.info(msg)
+    cc_logout_url = getattr(settings, 'CHAMELEON_PORTAL_SSO_BASE_URL', 'https://www.chameleoncloud.org') \
+        + getattr(settings, 'CHAMELEON_PORTAL_SSO_LOGOUT_PATH', '/logout/')
+    extra_context = {'cc_logout_url': cc_logout_url}
 
     """ Securely logs a user out. """
-    return django_auth_views.logout(request, template_name='auth/logout.html')
\ No newline at end of file
+    return django_auth_views.logout(request, template_name='auth/logout.html', extra_context=extra_context)
diff --git a/openstack_dashboard/settings.py b/openstack_dashboard/settings.py
index 925babb..e41d5b3 100644
--- a/openstack_dashboard/settings.py
+++ b/openstack_dashboard/settings.py
@@ -185,7 +185,7 @@ INSTALLED_APPS = [
 
 TEST_RUNNER = 'django_nose.NoseTestSuiteRunner'
 AUTHENTICATION_BACKENDS = ('openstack_auth.backend.KeystoneBackend',)
-AUTHENTICATION_URLS = ['openstack_dashboard.cc_web_sso_urls','openstack_auth.urls',]
+AUTHENTICATION_URLS = ['openstack_auth.urls',]
 AUTH_USER_MODEL = 'openstack_auth.User'
 MESSAGE_STORAGE = 'django.contrib.messages.storage.fallback.FallbackStorage'
 
-- 
1.8.3.1

